secadvisor/pkg/extend_gremlin.go

```diff
@@ -0,0 +1,131 @@
+// Use only single quotes in the gremlin expression
+// Need to define `Extend` field as a map[string]interface{} subfield of `transform`
+// Structure in yml file is as follows:
+//   transform:
+//     type: secadvisor
+//     secadvisor:
+//       extend:
+//         - VAR_NAME1=<gremlin expression with substitution strings>
+//         - VAR_NAME2=<gremlin expression with substitution strings>
+type extendGremlin struct {
+	gremlinClient    *client.GremlinQueryHelper
+	gremlinExprCache *cache.Cache
+	gremlinTemplates []*template.Template
+	newVars          []string
+	nTemplates       int
+}
+
+func NewExtendGremlin(cfg *viper.Viper) *extendGremlin {
+	gremlinClient := client.NewGremlinQueryHelper(core.CfgAuthOpts(cfg))
+	extendStrings := cfg.GetStringSlice(core.CfgRoot + "transform.secadvisor.extend")
+	var gremlinTemplates []*template.Template
+	var newVars []string
+	var nTemplates int
+	nTemplates = 0
+	gremlinTemplates = make([]*template.Template, len(extendStrings))
+	newVars = make([]string, len(extendStrings))
+	// parse the extension expressions and save them for future use
+	for _, ee := range extendStrings {
+		logging.GetLogger().Debugf("extend gremlin expression: %s", ee)
+		substrings := strings.SplitN(ee, "=", 2)
+		vv, err := template.New("extend_template").Parse(substrings[1])
+		if err != nil {
+			logging.GetLogger().Errorf("NewExtendGremlin: template error: %s", err)
+			continue
+		}
+		gremlinTemplates[nTemplates] = vv
+		newVars[nTemplates] = substrings[0]
+		nTemplates++
+	}
+	p := &extendGremlin{
+		gremlinClient:    gremlinClient,
+		gremlinExprCache: cache.New(10*time.Minute, 10*time.Minute),
+		gremlinTemplates: gremlinTemplates,
+		newVars:          newVars,
+		nTemplates:       nTemplates,
+	}
+	return p
+}
+
+// Extend: main function called to perform the extension substitutions
+// we expect the data of a single flow per call
+func (e *extendGremlin) Extend(in *SecurityAdvisorFlow) {
+	for i := 0; i < e.nTemplates; i++ {
+		var tpl bytes.Buffer
+		var err error
+		// apply the gremlin expression
+		err = e.gremlinTemplates[i].Execute(&tpl, in)
+		if err != nil {
+			continue
+		}
+		result := tpl.String()
+		// check if result is already in cache; if not, perform gremlin request
+		result2, ok := e.gremlinExprCache.Get(result)
+		if !ok {
+			// need to perform the gremlin query
+			result3, err := e.gremlinClient.Query(result)
+			if err != nil {
+				logging.GetLogger().Errorf("Extend: gremlin query = %s", result)
+				logging.GetLogger().Errorf("Extend: gremlin query error; err = %s", err)
+				continue
+			}
+			if len(result3) == 0 {
+				logging.GetLogger().Errorf("Extend: gremlin query error; result = %s", result)
+				continue
+			}
+			var result4 []string
+			if err = json.Unmarshal(result3, &result4); err != nil {
+				logging.GetLogger().Errorf("Extend: unmarshal error; err = %s", err)
+				continue
+			}
+			// add new field in map with the desired key name
+			if 0 == len(result4) {
+				continue
```

Comment:
empty result should also be cached (else a query yielding an empty result will be fired again and again and overload skydive analyzer)

Reply:
OK